### ISSUE TITLE
support for launching CloudABI programs

### ISSUE THREAD

**Issue by mheily** [mheily/jobd#24]
CloudABI is a very cool new security framework that uses a YAML configuration file to launch programs using a [cloudabi-run](https://github.com/NuxiNL/cloudabi-utils/blob/master/src/cloudabi-run/cloudabi-run.c) wrapper.  I haven't used CloudABI before, but there is some [documentation](https://nuxi.nl/doc/freebsd/)

It would be nice if relaunchd could launch CloudABI programs, and take care of binding to sockets and passing them to the program. Here's an example I found of setting up a CloudABI environment:

```
rm -Rf tmpdir
mkdir tmpdir
cloudabi-run _obj/unittest << EOF
%TAG ! tag:nuxi.nl,2015:cloudabi/

---
tmpdir: !file
  path: tmpdir
logfile: !fd stdout
nthreads: !!int 8
EOF
```

Here is a rough idea of what the above commands might look like as a relaunchd job:

```
{
"Label": "com.heily.cloudabi_test",
"Program": "/tmp/unittest",
"CloudABI": {
    "Tag": "nuxi.nl,2015:cloudabi/",
    "Threads": 8,
    "Tmpdir": {
         "Path": "/tmp",
    },
    "Logfile": "/dev/stdout",
}
```

@EdSchouten does this sound like something that would be a useful way to launch CloudABI daemons?

**Comment by EdSchouten** [comment]
Yes, that would be awesome! Just a couple of random thoughts:

The `cloudabi-run` man page contains a couple more examples of how to launch applications. There is even a tiny web server included that you can use to experiment.

`cloudabi-run` is just a tiny wrapper around libyaml and a (normally CloudABI specific) API for constructing the argument data and starting programs: [`<argdata.h>`](https://github.com/NuxiNL/cloudlibc/blob/master/src/include/argdata.h) and [`<program.h>`](https://github.com/NuxiNL/cloudlibc/blob/master/src/include/program.h). What's nice is that the cloudabi-utils package ships with a libcloudabi that provides a native port of these functions. All you need to do is invoke `argdata_*()` to construct the argument data and pass that on to `program_exec()`. The [`cloudabi-run` source code](https://github.com/NuxiNL/cloudabi-utils/blob/master/src/cloudabi-run/cloudabi-run.c) serves as a fine example of how that works.

I hope this information already helps. There's also [a recording of my talk at 32C3](https://www.youtube.com/watch?v=62cYMmSY2Dc) that should be well worth your time if you're interested in a more general overview.

**Comment by mheily** [comment]
Yep, that talk was helpful.. I also re-watched your interview on BSD Now.

There is actually a lot of similarity between what CloudABI needs and what I was planning to do for Capsicum, which is something like this:

```
{
  "Label": "mydaemon",
  "Program": "/usr/local/sbin/mydaemon",
  "User": "nobody",
  "Group": "nogroup",
  "Capsicum": {
    "CapabilitySet": {
      "kq_fd": {
        "SystemCall":["kqueue"], 
        "Rights": ["CAP_KQUEUE"]},
      },
      "resolv_fd":{
        "SystemCall": ["open", "/etc/resolv.conf", "O_RDONLY"], 
        "Rights": ["CAP_READ", "CAP_SYNC", "CAP_EVENT"],
      },
      "data_fd": {
        "SystemCall": ["open", "/var/mydata", "O_RDONLY"], 
        "Rights": ["CAP_CREATE"],
      },
    }
  }
}
```

In the above example, launchd is performing system calls on behalf of the program, and then applying cap_rights_limit() to restrict the capabilities of the file descriptors. It then will be calling cap_enter(), and executing the requested Program, so the program will inherit the file descriptors that it needs with the appropriate permissions already set.

So, given that CloudABI doesn't need to call cap_rights_limit() on it's descriptors, I think we can simplify the JSON to look more like this:

```
{
  "Label": "mydaemon",
  "Program": "/usr/local/sbin/mydaemon",
  "User": "nobody",
  "Group": "nogroup",
  "CloudABI": {
    "DescriptorSet": {
      "kq_fd": ["kqueue"], 
      "resolv_fd": ["open", "/etc/resolv.conf", "O_RDONLY"], 
      "data_fd": ["open", "/var/mydata", "O_RDONLY"],
    },
    "Options": {
       "key1": "value",
       "key2": [1, 2, 3],
    }
  }
}
```

For every element of the DescriptorSet, launchd will perform the requested system call, and map the fd to a name. For example, it will call kqueue() and assign that the name of "kq_fd".

Everything under the "Options" key would be passed into the program as configuration options. These would not be interpreted by launchd at all.

The tricky part will be managing the handoff between launchd and program_main(). One option is to have launchd emit YAML and pass it over stdin, but this does not seem very elegant. Perhaps I should look more closely at how liblaunch works, and see if it is worthwhile to add something like this to relaunchd.

**Comment by mheily** [comment]
liblaunch isn't bad, but I'm thinking that [nv(3)](https://www.freebsd.org/cgi/man.cgi?query=nv&apropos=0&sektion=3&manpath=FreeBSD+11-current&format=html) is the new hotness.
